sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java

```diff
@@ -188,6 +191,15 @@
         .build();
   }
 
+  /** Read Beam {@link Row}s from a JDBC data source. */
+  @Experimental(Experimental.Kind.SCHEMAS)
+  public static ReadRows readRows() {
+    return new AutoValue_JdbcIO_ReadRows.Builder()
+        .setFetchSize(DEFAULT_FETCH_SIZE)
+        .setOutputParallelization(true)
+        .build();
+  }
```

Comment:
This shouldn't be limited to just reading Row objects, as any type can have a schema. So you should be able to have a readWithSchema(Class<T> clazz) and get back a PCollection<T> (if the SchemaRegistry contains a schema for clazz, and if that schema matches the one inferred at the source). readRows is still useful but can be a wrapper around this.

Reply:
Do you prefer an explicit `readWithSchema` method or do you prefer it to be implicit?. That is:

- Make the JDBC row mapper optional
- If the user provides a JDBC row mapper, then:
    - If the output type of the row mapper exists in the registry, attach that schema to the output PCollection
    - Otherwise, the output PCollection will not have a schema
- If the user does not provide a JDBC row mapper, then:
    - If the output type exists in the schema registry and is compatible with the inferred schema from the database, generate our own row mapper to map the resultset and attach the schema to the output PCollection
    - Otherwise, throw an exception

Reply:
I think we can make it implicit in this case.

Reply:
I looked into this a little bit more and I think I may have misunderstood a few things.

I had the idea that if I get the `TypeDescriptor` of the PCollection, I could query the schema registry to get the necessary information to attach the schema. However, this doesn't seem to work in practice. Of course, I could be doing something wrong there but I came to the realisation that if it was indeed a viable solution, then it would have been already implemented at the core as it would provide a generic way to implicitly attach schemas to any PCollection that contains suitable element types.

Now I am not quite sure what benefit a `readWithSchema(clazz)` method would have. If the user does need schema support, they could just as well call `setSchema` on the PCollection themselves, can't they? We can reduce the boilerplate involved but that is not specific to the JDBC IO module and should probably go in to the core schema module as a utility function.

Since the user is responsible for defining how to map the ResultSet to a type of their choosing, they could potentially do some field transforms during the mapping and produce an output type whose schema doesn't match up with the schema inferred from the raw data. Therefore I feel like comparing the two schemas will not be very useful either.

What are your thoughts?

Reply:
I was thinking that this would allow us to not have to create a separate fork of the Read PTransform. Just keep one Read<T>, which happens to support schemas (as long as T is in the SchemaRegistry); yes the user can call setSchema themselves, however forcing the user to do a lookup in the SchemaRegistry themselves in ugly, and we should make that implicit. Then readRows will just return Read<Row>, so we should not need a separate class.

Does that work?

Reply:
So I assumed that it would be easy to obtain the schema for `T` and implicitly attach it to the PCollection. However, this appears to be difficult to do reliably unless the user explicitly provides a reference to `Class<T>` as you correctly suggested. 

Given that the user has to make a conscious effort to enable schema support by providing a reference to `Class<T>`, wouldn't it make sense to introduce a generic utility function such as `PCollections.withSchema(PCollection<T>, Class<T>)` that would work for any kind of PCollection and not just the output of JDBC IO?